### ISSUE TITLE
`asyncdispatch` exception handling example does not work as stated

### ISSUE THREAD

**Issue by btcross26** [nim-lang/Nim#17703]
I am unable to capture an exception when using async procedures when using Nim 1.4.2 or Nim 1.4.4.

TLDR: I am trying to set up a simple server that takes a POST request and does some stuff with the data using some C dynamic library functions. No matter what I do, I am unable to catch exceptions using `asynchttpserver`. At first, I figured this could have something to do with the C aspect of this, but the below toy example fails to work also.

***Below toy example is for Nim 1.4.4 with the `asyncdispatch` advice on exception handling. I have not been able to get anything to work with try-except either.***

### Example
```nim
proc cb(req: Request): Future[void] {.async.} =
  raise newException(ValueError, "this is a dummy error")

proc main() {.async.}=
  var
    future: Future[void]
    server = newAsyncHttpServer()
  server.listen(Port(8080))
  while true:
    if server.shouldAcceptRequest():
      future = server.acceptRequest(cb)
      echo "You are here!"
      yield future
      if future.failed:
        echo "Future dog!"
      elif future.finished:
        echo "Future cat!"
    else:
      poll()

waitFor main()
```

### Current Output
If I compile the above, run the server, and hit it with a simple POST request from curl, I get a `curl: (52) Empty reply from server` on the request side, and the below failure from the server:

```
You are here!
Future cat!
You are here!
....[lots of traceback]
Error: unhandled exception: this is a dummy error
Async traceback:
  /Users/ben/Documents/github/nim_demo/model_server.nim(114)                   model_server
  /Users/ben/.choosenim/toolchains/nim-1.4.4/lib/pure/asyncdispatch.nim(1935)  waitFor
  /Users/ben/.choosenim/toolchains/nim-1.4.4/lib/pure/asyncdispatch.nim(1627)  poll
  /Users/ben/.choosenim/toolchains/nim-1.4.4/lib/pure/asyncdispatch.nim(1368)  runOnce
  /Users/ben/.choosenim/toolchains/nim-1.4.4/lib/pure/asyncdispatch.nim(208)   processPendingCallbacks
  /Users/ben/.choosenim/toolchains/nim-1.4.4/lib/pure/asyncmacro.nim(32)       processRequestNimAsyncContinue
  /Users/ben/.choosenim/toolchains/nim-1.4.4/lib/pure/asynchttpserver.nim(270) processRequestIter
  /Users/ben/.choosenim/toolchains/nim-1.4.4/lib/pure/asyncmacro.nim(262)      cb
  /Users/ben/.choosenim/toolchains/nim-1.4.4/lib/pure/asyncmacro.nim(29)       cbNimAsyncContinue
  /Users/ben/Documents/github/nim_demo/model_server.nim(95)                    cbIter
  #[
    /Users/ben/Documents/github/nim_demo/model_server.nim(114)                   model_server
    /Users/ben/.choosenim/toolchains/nim-1.4.4/lib/pure/asyncdispatch.nim(1935)  waitFor
    /Users/ben/.choosenim/toolchains/nim-1.4.4/lib/pure/asyncdispatch.nim(1627)  poll
    /Users/ben/.choosenim/toolchains/nim-1.4.4/lib/pure/asyncdispatch.nim(1368)  runOnce
    /Users/ben/.choosenim/toolchains/nim-1.4.4/lib/pure/asyncdispatch.nim(208)   processPendingCallbacks
    /Users/ben/.choosenim/toolchains/nim-1.4.4/lib/pure/asyncmacro.nim(32)       processRequestNimAsyncContinue
    /Users/ben/.choosenim/toolchains/nim-1.4.4/lib/pure/asyncmacro.nim(145)      processRequestIter
    /Users/ben/.choosenim/toolchains/nim-1.4.4/lib/pure/asyncfutures.nim(372)    read
...[more traceback]
Exception message: this is a dummy error
Exception type: [ValueError]
```

### Expected Output
As per the example, I would expect to land in the future failed block.
```
You are here!
Future Dog!
[other stuff and maybe a potential failure]
```

### Possible Solution

* Unsure, but it seems as though the `error` attribute of the future is not being set. I have tried other means of doing this by explicitly calling the `fail(future, error)` proc, but have been unable to succeed. Also, notice from the output that it appears to get raised on the second `yield` call.

### Additional Information
My actual server example works and responds correctly with input that won't trigger an exception - i.e., it processes the POST data, calls my C functions, and responds with data. The server is simply serving an XGBoost model. It expects a POST request with JSON of inputs, scores the model, and returns the score. As the model itself is a meaningless, toy example, it expects 10 inputs. If I give it less, it will raise an exception and I would like to respond accordingly in the data I return.

The curl simple POST request to trigger above behavior:
```
curl -X POST --data '{"f1": 0.5010}' http://localhost:8080
```

```
$ nim -v
Nim Compiler Version 1.4.4 [MacOSX: amd64]
Compiled at 2021-02-27
Copyright (c) 2006-2020 by Andreas Rumpf

active boot switches: -d:release
```

**Comment by btcross26** [comment]
# Resolution

Turns out in the end, that it WAS the inability to capture the exception from C code, rather than Nim. A few things to note:

1. As it stands, there is an `AsyncCheck` within the `acceptRequest` method shown in main above that prevents error handling from making it back to `proc` main. If you change this to `await` instead, the server will continue to run and raise the error back to `main` (showing "Future Dog"), but the client side will just hang since the connection was not closed and no response is given. This is because the callback (`cb`) gets called in isolation well down the line in a proc called `processRequest` before the connection is closed - there is no way to capture the error there without rewriting the call chain. This maybe could be handled within the accept request method with a second "on-error" callback, but not worth changing.

2. As written in the first post, a try-except clause would capture the error shown. However, the try-except clause would not capture the error from the C library (although it would properly display the message for the unhandled exception). I probably was too wishful here in compiling Nim code with Nim exceptions into a C library and hoping I could catch the exceptions.

3. The easiest way around was to change the C dynamic library function in a way that returned a status message reflecting whatever exception I was previously trying to catch, and returned a score by modifying a float variable that was passed to the function by reference. So the dynamic library function was changed to something like this:

```nim
proc modelPredict(jsonInputs: cstring, score: ptr cfloat): cstring =
  var
    dmatrix: DMatrixHandle = ptrInitializer
    output_length: bst_ulong
    output_result: ptr cfloat = cast[ptr cfloat](ptrInitializer)
    inputs: seq[cfloat]
  # score model
  try:
    inputs = jsonToFeatures(jsonInputs)
  except:
    return "input json error"

  try:
    safe_xgboost: XGDMatrixCreateFromMat(cast[ptr cfloat](inputs[0].unsafeAddr()),
      1, bst_ulong(inputs.len()), missing_value, dmatrix.addr())
    safe_xgboost: XGBoosterPredict(booster, dmatrix, 0, 0, 0, output_length.addr(),
        output_result.addr())
    safe_xgboost: XGDMatrixFree(dmatrix)
    score[] = output_result[]
  except:
    return "XGBError: $1" % [$XGBGetLastError()]

  return "OK"
```

And the server changed to something like this:

```nim
proc cb(req: Request): Future[void] {.async.} =
  let headers = {
    "Date": $now(),
    "Content-type": "application/json"
  }
  var
    code: HttpCode
    responseHeaders: HttpHeaders = newHttpHeaders(headers)
    outputJson: JsonNode
    score: float32
    message: string
  outputJson = %* {"score": newJNull()}
  if req.reqMethod == HttpPost:
    message = $modelPredict(cstring(req.body), score)
    outputJson["status"] = %message
    if message == "OK":
      outputJson["score"] = %score
      code = Http200
    elif message == "input json error":
      code = Http400
    else:
        code = Http500
  else:
    outputJson["status"] = %"invalid request type"
    code = Http405
  await req.respond(code, $outputJson, responseHeaders)

proc main(port: Port) {.async.}=
  var
    server = newAsyncHttpServer()
  server.listen(port)
  while true:
    if server.shouldAcceptRequest():
      await server.acceptRequest(port, cb)
    else:
      poll()

when isMainModule:
  var port: Port
  if paramCount() == 0:
    port = Port(8080)
  else:
    port = Port(paramStr(1).parseInt())
  loadModel()
  waitFor main(port)
  freeModel() 
```
